File: Reverse Linked List/Leetcode_206.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */

// iterative

class Solution {
public:
    ListNode* reverseList(ListNode* head) {
        
        ListNode* cur = head; ListNode* prev = NULL; ListNode* next = NULL; 
        
        while (cur != NULL)
        {
            next = cur->next;
            cur->next = prev;
            prev = cur;
            cur = next;
        }
        
        return prev;
    }
};

// recursive
class Solution {
public:
    ListNode* reverseList(ListNode* head) {
        
        if (!head || !(head->next))
        {
            return head;
        }
        
        ListNode* tmp = reverseList(head->next);
        head -> next -> next = head;
        head -> next = NULL;
        return tmp;
    }
};

File: Reverse Linked List/Leetcode_206.py
#Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:
    def reverseList(self, head: ListNode) -> ListNode:
        cur, prev, _next = head, None, None
        while cur:
            _next = cur.next
            cur.next = prev
            cur, prev = _next, cur
        return prev
            

# recursive solution
class Solution:
    def reverseList(self, head: ListNode) -> ListNode:
        if head is None or head.next is None: return head

        dummy = self.reverseList(head.next)
        head.next.next = head
        head.next = None
        return dummy
